src/Client/LocalConnection.cpp

```diff
@@ -0,0 +1,411 @@
+    }
+
+    if (state->block && state->block.value())
+    {
+        next_packet_type = Protocol::Server::Data;
+        return true;
+    }
+
+    return false;
+}
+
+bool LocalConnection::pollImpl()
+{
+    Block block;
+    auto next_read = pullBlock(block);
+    if (block)
+    {
+        state->block.emplace(block);
+    }
+    else if (!next_read)
+    {
+        state->is_finished = true;
+    }
+
+    return true;
+}
+
+Packet LocalConnection::receivePacket()
+{
+    Packet packet;
+    if (!state)
+    {
+        packet.type = Protocol::Server::EndOfStream;
+        return packet;
+    }
+
+    if (!next_packet_type)
+        poll();
+
+    if (!next_packet_type)
+    {
+        packet.type = Protocol::Server::EndOfStream;
+        return packet;
+    }
+
+    packet.type = next_packet_type.value();
+    switch (next_packet_type.value())
+    {
+        case Protocol::Server::Totals: [[fallthrough]];
+        case Protocol::Server::Extremes: [[fallthrough]];
+        case Protocol::Server::Log: [[fallthrough]];
+        case Protocol::Server::Data:
+        {
+            if (state->block && state->block.value())
+            {
+                packet.block = std::move(state->block.value());
+                state->block.reset();
+            }
+            break;
+        }
+        case Protocol::Server::Exception:
+        {
+            packet.exception = std::make_unique<Exception>(*state->exception);
+            break;
+        }
+        case Protocol::Server::Progress:
+        {
+            packet.progress = std::move(state->progress);
+            state->progress.reset();
+            break;
+        }
+        case Protocol::Server::EndOfStream:
+        {
+            break;
+        }
+        default:
+            throw Exception("Unknown packet " + toString(packet.type)
+                + " from server " + getDescription(), ErrorCodes::UNKNOWN_PACKET_FROM_SERVER);
+    }
+
+    next_packet_type.reset();
+    return packet;
+}
+
+void LocalConnection::getServerVersion(
+    const ConnectionTimeouts & /* timeouts */, String & /* name */,
+    UInt64 & /* version_major */, UInt64 & /* version_minor */,
+    UInt64 & /* version_patch */, UInt64 & /* revision */)
+{
+    throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Not implemented");
+}
+
+void LocalConnection::setDefaultDatabase(const String &)
+{
+    throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Not implemented");
+}
+
+UInt64 LocalConnection::getServerRevision(const ConnectionTimeouts &)
+{
+    throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Not implemented");
+}
+
+const String & LocalConnection::getServerTimezone(const ConnectionTimeouts &)
+{
+    throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Not implemented");
+}
+
+const String & LocalConnection::getServerDisplayName(const ConnectionTimeouts &)
+{
+    throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Not implemented");
+}
+
+void LocalConnection::sendExternalTablesData(ExternalTablesData &)
+{
+    /// Do nothing.
```

Comment:
Why?

Reply:
Because for client it must be called even if external tables vector size is zero, so I think it is the nicest solution.

Reply:
It's a bit confusing, because external tables silently do not work:
```
$ echo -e "1\n2\n3" > file.tsv
$ ./clickhouse client --query="SELECT * FROM data" --external --file=file.tsv --name=data --types=UInt32
1
2
3
$ ./clickhouse local --query="SELECT * FROM data" --external --file=file.tsv --name=data --types=UInt32
Code: 60. DB::Exception: Table _local.data doesn't exist. (UNKNOWN_TABLE) (version 21.11.1.8120)
```

Error message looks like table name is specified incorrectly, so user may spent some time trying to understand what they are doing wrong. The only way to find out that the feature is actually not implemented is to read your comment on github (the comment in code does not really help, because "Do nothing" may sound like "There is nothing to do, external tables are already attached to current session context").

Reply:
Ok, I'll add a `supports_external_tables` flag then and throw if needed.

Reply:
Another way is to implement `sendExternalTablesData(...)`, but a flag and correct error message are ok for now too.